synapse/logging/tracing.py

```diff
@@ -477,211 +457,196 @@ def whitelisted_homeserver(destination: str) -> bool:
+    kind: Optional["opentelemetry.trace.SpanKind"] = SpanKind.INTERNAL,
+    attributes: "opentelemetry.util.types.Attributes" = None,
+    links: Optional[Sequence["opentelemetry.trace.Link"]] = None,
+    start_time: Optional[int] = None,
+    record_exception: bool = True,
+    set_status_on_exception: bool = True,
+    end_on_exit: bool = True,
+    # For testing only
+    tracer: Optional["opentelemetry.trace.Tracer"] = None,
+) -> ContextManager["opentelemetry.trace.span.Span"]:
+    if opentelemetry is None:
         return contextlib.nullcontext()  # type: ignore[unreachable]
 
-    references = [opentracing.follows_from(context) for context in contexts]
-    scope = start_active_span(
-        operation_name,
-        child_of=child_of,
-        references=references,
+    # TODO: Why is this necessary to satisfy this error? It has a default?
+    #  ` error: Argument "kind" to "start_span" of "Tracer" has incompatible type "Optional[SpanKind]"; expected "SpanKind"  [arg-type]`
+    if kind is None:
+        kind = SpanKind.INTERNAL
+
+    span = start_span(
+        name=name,
+        context=context,
+        kind=kind,
+        attributes=attributes,
+        links=links,
         start_time=start_time,
-        tracer=tracer,
+        record_exception=record_exception,
+        set_status_on_exception=set_status_on_exception,
     )
 
-    if inherit_force_tracing and any(
-        is_context_forced_tracing(ctx) for ctx in contexts
-    ):
-        force_tracing(scope.span)
-
-    return scope
+    # Equivalent to `tracer.start_as_current_span`
+    return opentelemetry.trace.use_span(
+        span,
+        end_on_exit=end_on_exit,
+        record_exception=record_exception,
+        set_status_on_exception=set_status_on_exception,
+    )
 
 
 def start_active_span_from_edu(
-    edu_content: Dict[str, Any],
     operation_name: str,
-    references: Optional[List["opentracing.Reference"]] = None,
-    tags: Optional[Dict[str, str]] = None,
-    start_time: Optional[float] = None,
-    ignore_active_span: bool = False,
-    finish_on_close: bool = True,
-) -> "opentracing.Scope":
+    *,
+    edu_content: Dict[str, Any],
+) -> ContextManager["opentelemetry.trace.span.Span"]:
     """
     Extracts a span context from an edu and uses it to start a new active span
 
     Args:
+        operation_name: The label for the chunk of time used to process the given edu.
         edu_content: an edu_content with a `context` field whose value is
-        canonical json for a dict which contains opentracing information.
-
-        For the other args see opentracing.tracer
+            canonical json for a dict which contains opentracing information.
     """
-    references = references or []
-
-    if opentracing is None:
+    if opentelemetry is None:
         return contextlib.nullcontext()  # type: ignore[unreachable]
 
     carrier = json_decoder.decode(edu_content.get("context", "{}")).get(
-        "opentracing", {}
-    )
```

Comment:
> AFAICT, this never worked because this PR serializes everything into `content["org.matrix.opentracing_context"]` but `start_active_span_from_edu` reads from `content["opentracing"]` (even at this point in time in the code to current day).
> 
> https://github.com/matrix-org/synapse/blob/bbc23cdd04a2dfafefcbcb3264c2e89b50eacc04/synapse/logging/opentracing.py#L402
> 
> *-- @MadLittleMods, https://github.com/matrix-org/synapse/pull/5852#discussion_r934960586*

Reply:
Do we get value out of this trace connection from the homeserver EDU sender to the homeserver that receives the EDU? Wondering about ripping it all out

Reply:
Moved to https://github.com/matrix-org/synapse/pull/13400#discussion_r934978867